refactor: Material mapping howto for python bindings (#1801)
acts-project/acts: docs/howto/run_material_mapping.rst

Mapping and configuration
-------------------------

> First we need to extract the list of all the surfaces and volumes in our detector, to do so we will use the ``geometry.py`` script :

Reviewer: ```suggestion
First we need to extract the list of all the surfaces and volumes in our detector. To do so we will use the ``geometry.py`` script:
```

.. code-block::

   $ python3 <source>/Examples/Scripts/Python/geometry.py 

> Ideally the following options should be used in the python file :

Reviewer: ```suggestion
Ideally the following options should be used in the python file:
```

.. code-block::

   def runGeometry(
      trackingGeometry,
      decorators,
      outputDir,
      events=1,
      outputObj=False,
      outputCsv=False,
      outputJson=True,
      outputRoot=False,
   ):

> For the following example we will be remapping the material of the ODD, we will thus get our detector via the following line :

Reviewer: ```suggestion
For the following example we will be remapping the material of the ODD, we will thus get our detector via the following line:
```

----------------

With the surfaces map and the material track we can finally do the material mapping using the ``material_mapping.py`` script:

.. code-block:: console

   $ python3 <source>/Examples/Scripts/Python/material_mapping.py 

Note that technically when using DD4hep (in particular for the ODD) defining a ``matDeco`` in the main function is not strictly necessary as the DD4hep geometry can hold the information of which surface to map onto with which binning. We will ignore this option, since the goal of this guide is to explain how to make a material map regardless of the detector.

As an output you will obtain the material map as a root and JSON file and a new material track collection in a root file. This new collection adds to each material interaction the associated surface during the mapping. This can be used for the control plots.
Depending on what you want to do there are three options you can change:

- ``mapSurface``: determine if material is mapped onto surfaces
- ``mapVolume``: determine if material is mapped onto volumes
- ``mappingStep``: determine the step size used in the sampling of the volume. This should be small compared to the bin size.

> - ``readCachedSurfaceInformation`` if added the material-surface association will be taken from the input material track file (doesn't work with geantino file, you need to use the material track file obtained from running the material mapping).

Reviewer 2: Considering you are mentioning the options used in the material mapping, can you also document here what the `mappingStep` refers to?

Author: Sure I can go into a bit more detail

Author: Does this make sense to you : 

- ``mappingStep``: determine the step size used in the sampling of the volume. The material interaction point obtained though G4 are only at the first intersection between the track and the material, this means that the mapping will be incorrect if the material is deeper than the bin size. To avoid this, we create new additional material point spaced equally every ``mappingStep`` along the trajectory. This should be small compared to the bin size.

Reviewer 2: Does this parameter then make sense only for mapping onto volumes?
This is my suggestion, feel free to edit it further:
By default, the material interaction point obtained from G4 is accumulated at the intersection between the track and the volume material. The mapping will be therefore incorrect if the material extends through the bin. To avoid this, additional material points are created every mappingStep [mm] along the trajectory. The mapping step should be small compared to the bin size.

Author: Exactly it is only used in the volume mapping. The surface mapper doesn't use this parameters. Thanks for the  suggestion !

Author: I have updated the doc. It seems that now we have an issue in the CI related to the codding guideline page.

Using a different detector
--------------------------

> If you want to use a different type of detector, you will first need to ensure that the relevant packages were added during the compilation. After this you can just replace the detector initialisation in the different main function. For reference you can base your self on the Odd for DD4Hep detector and on the ITk for TGeo detector. 

Reviewer: ```suggestion
If you want to use a different type of detector, you will first need to ensure that the relevant packages were added during the compilation. After this you can just replace the detector initialisation in the different main function. For reference you can have a look on the Odd for DD4Hep detector and on the ITk for TGeo detector. 
```
